### PR TITLE
Speed up BalancedShardAllocator a Little

### PR DESCRIPTION
**Pull request by original-brownbear**
`addShard` is quite hot when profiling in many-shards benchmarks.
This saves one lookup in the hot loop.
Also fixed the confusing naming in `ModelIndex`.

relates #77466

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java [hunk 1]
```diff
@@ -1172,12 +1167,12 @@ public boolean containsShard(ShardRouting shard) {
     }
 
     static final class ModelIndex implements Iterable<ShardRouting> {
-        private final String id;
+        private final String indexName;
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java:1170]: Is this field used for anything?

**Review by idegtiarenko — APPROVED** [review]

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java:1170]: It was used here:

```
                indices.put(index.getIndexId(), index);
```

That wasn't necessary tho, it's the same as `shard.getIndexName()` which was already in scope. I imagine it was useful for development/debugging in the past but no need for it any more IMO.

**Review by DaveCTurner — APPROVED** [review]
LGTM2 but I agree with Ievgen about dropping the field.

**Comment by original-brownbear** [comment]
Thanks Ievgen + David!

**Pull request merged**